### PR TITLE
Move to new rabbit client 4.x

### PR DESCRIPTION
**Pull request by igreenfield**
Add new parameters for NIO.
Expose statistics with JMX

### CODE REVIEW

File: queue-rabbitmq/src/main/java/com/cisco/oss/foundation/message/RabbitMQMessagingFactory.java [hunk 1]
```diff
@@ -267,7 +275,57 @@ public void onRecoveryFailure(Consumer consumer, Channel channel, Throwable fail
             ConnectionOptions options = new ConnectionOptions()
                     .withAddresses(addresses.toArray(addrs));
 
-            Configuration configuration = ConfigurationFactory.getConfiguration();
+            final boolean metricsAndMonitoringIsEnabled = configuration.getBoolean("service.rabbitmq.MetricsAndMonitoringJmx.isEnabled", false);
```
Comment (yairogen) [inline queue-rabbitmq/src/main/java/com/cisco/oss/foundation/message/RabbitMQMessagingFactory.java]: change MetricsAndMonitoringJmx to camel case

File: queue-rabbitmq/src/main/resources/configSchema.xml [hunk 2]
```diff
@@ -141,6 +141,20 @@
...
         </Parameter>
 
+        <Parameter name="service.rabbitmq.useNio" type="BOOLEAN"
+                   description="Use Java Nio while connecting to RabbitMQ">
+            <DefaultValue>
+                <PrimitiveValue value="false"/>
+            </DefaultValue>
+        </Parameter>
+
+        <Parameter name="service.rabbitmq.MetricsAndMonitoringJmx.isEnabled" type="BOOLEAN"
```
Comment (yairogen) [inline queue-rabbitmq/src/main/resources/configSchema.xml]: update name as per above.

File: queue-rabbitmq/src/main/java/com/cisco/oss/foundation/message/RabbitMQMessagingFactory.java [hunk 3]
```diff
@@ -267,7 +275,57 @@ public void onRecoveryFailure(Consumer consumer, Channel channel, Throwable fail
...
+                JmxReporter reporter = JmxReporter
+                        .forRegistry(registry)
+                        .inDomain("com.rabbitmq.client.jmx")
+                        .build();
+                reporter.start();
+            }
+
+            final boolean useNio = configuration.getBoolean("service.rabbitmq.useNio", false);
+
+            if (useNio) {
```
Comment (yairogen) [inline queue-rabbitmq/src/main/java/com/cisco/oss/foundation/message/RabbitMQMessagingFactory.java]: does the rest of the flow stay the same when Nio is enabled or does the connect api need to be used differently? in other words do we need an else for this 'if'?

Comment (igreenfield) [inline queue-rabbitmq/src/main/java/com/cisco/oss/foundation/message/RabbitMQMessagingFactory.java]: all stay the same

**Comment by igreenfield** [comment]
@yairogen Please rereview

**Comment by yairogen** [comment]
did you tests all recovery scenarios in both NIO and regular mode?

**Comment by igreenfield** [comment]
Yes

[banner1]



Izek Greenfield
ENGINEER.SOFTWARE ENGINEERING
igreenfi@cisco.com<mailto:igreenfi@cisco.com>
Tel: +972 2 588 6286

Cisco Systems, Inc.
5 Shlomo (Momo) Halevi Street Har Hotzvim High Tech Park
JERUSALEM
9777019
Israel
cisco.com


[http://www.cisco.com/assets/swa/img/thinkbeforeyouprint.gif]Think before you print.

This email may contain confidential and privileged material for the sole use of the intended recipient. Any review, use, distribution or disclosure by others is strictly prohibited. If you are not the intended recipient (or authorized to receive for the recipient), please contact the sender by reply email and delete all copies of this message.
Please click here<http://www.cisco.com/web/about/doing_business/legal/cri/index.html> for Company Registration Information.

**Pull request merged**